### PR TITLE
replicate: implement recover progress

### LINKED ISSUE #2273: Synchronous replication: recover progress
In order to switch from `sync-recover` to `sync` mode. PD needs to make sure all region's states become `integrity-over-label` and show the progress in real time.

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
Fix #2273

### What is changed and how it works?
- save region replicate status to cache
- calculate recover progress based on region status (a simple approach, need improve, see: https://github.com/pingcap/pd/issues/2327)
- update test

### Check List 

Tests 
 - Unit test

Code changes
 - Has persistent data change

### CODE REVIEW

File: server/replicate/replicate.go [hunk 1]
```diff
@@ -227,13 +231,35 @@ func (m *ModeManager) checkCanSync() bool {
...
+			return
+		}
+
+		total += len(regions)
+		for _, r := range regions {
+			if !bytes.Equal(key, r.GetStartKey()) {
+				log.Warn("found region gap", zap.ByteString("key", key), zap.ByteString("region-start-key", r.GetStartKey()))
+				total++
+			}
+			if r.GetReplicateStatus().GetRecoverId() == m.drAutosync.RecoverID &&
```
Comment (nolouch) [inline server/replicate/replicate.go:251]: StateID?

Comment (disksing) [reply server/replicate/replicate.go:251]: Update it in later PR. The kvproto is not updated here.

File: server/replicate/replicate.go [hunk 2]
```diff
@@ -227,13 +231,35 @@ func (m *ModeManager) checkCanSync() bool {
...
+		if len(regions) == 0 {
+			log.Warn("scan empty regions", zap.ByteString("start-key", key))
+			total++ // make sure it won't complete
+			return
+		}
+
+		total += len(regions)
+		for _, r := range regions {
+			if !bytes.Equal(key, r.GetStartKey()) {
+				log.Warn("found region gap", zap.ByteString("key", key), zap.ByteString("region-start-key", r.GetStartKey()))
```
Comment (rleungx) [inline server/replicate/replicate.go:248]: How about also printing the region ID?

Comment (disksing) [reply server/replicate/replicate.go:248]: ok

File: server/replicate/replicate.go (hunk 1 above)
Comment (AndrewDi) [reply server/replicate/replicate.go:251]: @disksing Do we need to check region state, invalid region should not include in `total` count?

**Pull request merged**